### ISSUE TITLE
Updates: extension version only updated on current site when not network active

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#706]
In https://github.com/WordPoints/wordpoints/issues/404#issuecomment-314846091 and https://github.com/WordPoints/wordpoints/issues/404#issuecomment-314848681 we explain how when WordPoints is network active, a per-site activated extension will only have its version number updated on the site that the update is running from, because this isn't automatically done as part of the un/installer's built-in bootstrap for the update routine.

This is a problem, because as pointed out in https://github.com/WordPoints/wordpoints/issues/404#issuecomment-314560565 we run the update routine for all sites where the extension is installed at once, so the other sites will actually be up-to-date after the update routine is run from any one of them.

One result is that the update routine will be repeated for every site where the entity is active, which will probably lead to breakage since the updates are not designed with idempotence in mind.

**Comment by JDGrimes** [comment]
Note that both in the current logic and the new logic (https://github.com/WordPoints/wordpoints/issues/404#issuecomment-314860025) there is an additional issue, in that we skip the running of the update routines for each site if there is no per-site updates to run (since naturally it would just be a waste to switch to every site on the network). That also needs to be taken into account in fixing this issue; depending on the solution we may have to switch to each site only to update the version number.

**Comment by JDGrimes** [comment]
I had thought that we had intentionally designed the update routines to run on all sites at once even when not network-installed because of `switch_to_blog()`. However, I just stumbled across https://github.com/WordPoints/wordpoints/issues/68#issuecomment-45744901, which shows that we had that thinking for *install*, and also that at least at that time we did *not* think that we were doing the same for updates. So we may want to rethink that approach, because as far as I can see that *is* what the current code is doing.

**Comment by JDGrimes** [comment]
It looks like the code for `update()` was correct to the behavior we then described when it was introduced in https://github.com/WordPoints/wordpoints/commit/e9f476570917b1ff150c929c9bde005cf85baf6f.

We always [defaulted to `is_wordpoints_network_active()`](https://github.com/WordPoints/wordpoints/commit/e9f476570917b1ff150c929c9bde005cf85baf6f#diff-3703b656220df5ad3d5d0607d6e9f005R184), which was perfectly fine for WordPoints and its components. Thus in https://github.com/WordPoints/wordpoints/commit/5f893845d38dda6f0b3d93f141355ee361a43614, for example, we did not pass in the `$network` parameter. This continued in https://github.com/WordPoints/wordpoints/commit/614c402459ac25b9bd6e253706b9fa6677a3b59c.

Where things fell apart was when we then began using the updater for modules, specifically, when we [factored out the installables class](https://github.com/WordPoints/wordpoints/commit/6bb33a96e8b07fb361b528c318769cac486e0fdb#diff-7ab4aa7ce8b16a55c1287f74a85e00d4R211), whose design was based on the code for core and its components (and so didn't pass the `$network` parameter), but was meant to be used for extensions as well. (Prior to that, the module had to call the `update()` method manually, so it was up to it to see that it passed in the correct params.)

**Comment by JDGrimes** [comment]
So, I guess we've identified that the real bug here is the fact that we aren't passing in the `$network` value and thus aren't running the updater only for the current site when per-site activated.

As far as the thoughts about `switch_to_blog()`, these concerns aren't really valid when per-site active, because each site is really independent at that point. WordPoints shouldn't be expected to properly communicate with itself across them or something. (Kind of related: #656).

**Comment by JDGrimes** [comment]
The `switch_to_blog()` concerns carry slightly more weight when we think about the case where WordPoints is network-activated, but an extension is only per-site activated. Still, I think that we can rightly call any problems relating to updates not happening before the switch an edge-case. Especially since we don't really anticipate an extension design that would cause issues in that case to start with, for the same reason that it should be isolated across sites then.

Though on the other hand, any time that we create global tables (as in WordPoints) we aren't truly isolated across sites. Perhaps that in itself is a bug though. And regardless, the table is then updated across all sites, just as the code is. So nothing would break because of that, unless the code checked the database version before determining how to use the DB. But that shouldn't be necessary, since updates are automatic like this.

In general, the database version should probably not be taken into account at all then, within the code. Only the code version.

**Comment by JDGrimes** [comment]
Wouldn't the network-level update routine get run multiple times then though? It would be triggered again for each site. Unless we specified for the network part not to run at all then. But it *does* have to run, it just only needs to run once.

I suppose as we said above, if an entity is installed per-site, it really shouldn't have a network install routine. Yet there are things that are global by nature, like the filesystem, or some database tables. Though those can still be used to store information per-site. This is a bigger issue than we can tackle right now; even if we are doing it wrong we still need to support it for the time being. So we'll need to properly handle per-site installed entities that have network install routines.

**Comment by JDGrimes** [comment]
As far as not passing the network status into the updater, we ended up fixing that as part of #404, because we redid the installables class. So we tell our new updater whether we are network wide, and thus it now (correctly) only runs the update per site, as it is triggered on each site, and the versions are updated for each site accordingly. 

However, there is a catch as it relates to legacy code, because entities using the old un/installers are now also triggered to update through the new installables app, via a shim installable class. But that legacy installable doesn't pass in the network information when it runs the un/installer's `update()` method. So that still needs to be fixed.

**Comment by JDGrimes** [comment]
> Note that both in the current logic and the new logic (#404 (comment)) there is an additional issue, in that we skip the running of the update routines for each site if there is no per-site updates to run (since naturally it would just be a waste to switch to every site on the network). That also needs to be taken into account in fixing this issue; depending on the solution we may have to switch to each site only to update the version number.

Note that this is really irrelevant, because we're only dealing with per-site activated installables here, not network-active ones.

**Comment by JDGrimes** [comment]
I don't think that it is really necessary to take action here to fix the non-updated database versions. Actually, they probably are all up-to-date already anyway, because the other sites will have loaded, and the update will have been triggered again on them. As such, if any issues were going to occur from this, they probably already would have.

The main thing left to fix here then is just the network updates running multiple times.

**Comment by JDGrimes** [comment]
Of course, that too is already happening for the same reasons as the above, and hasn't (apparently) caused any issues. But we probably don't want to enshrine it as a "feature" of the new API.

**Comment by JDGrimes** [comment]
I think we need to keep track of the version at both the network and per-site levels, and only run the network update if network version is out of date. However, that will clash with the current design of the installable's `get_update_routines()`, in which the version to update from is determined internally based on the activation status of the installable. In this case, we need to have the version be either the network or site version not based on activation status, but either version to be returned as determined is needed by the updater.

A simple solution would be to let whether the network version should be used be passed in as a boolean arg. Or, we could just have the version itself be passed in, instead.

Passing in the version seems like unnecessarily flexibility, but the issue I see with passing in just a network boolean is that it would be likely to be interpreted as indicating whether the entity was network active. It won't always correspond with that of course, as in the situation which we are particularly concerned with, so any logic based on it would be wrong.

Looking at it from a slightly different angle though, when we are getting the network routines, we only need the network routines, and when we are getting the site routines, we only need the site routines (in this case, that is; when network active we'd need both). So if the boolean was passed in, we could use it to return only the routines needed. Though on the other hand, we don't take whether we're on multisite into account. But maybe we should.

**Comment by JDGrimes** [comment]
> when we are getting the network routines, we only need the network routines, and when we are getting the site routines, we only need the site routines

Which is basically another way of saying that we do need them both, at once, but that we were thinking of getting them separately. By that wouldn't be necessary if we weren't passing in a boolean that just meant which version to use. That is, if we internally used the correct version for each update routine, we could return just the network and per-site ones needed.

Of course, that would require more logic within the update routines getter method, which we'd like to avoid. It means that essentially we're decentralizing and duplicating this logic. So that we are anticipating the operation of the updater from within the installable, making the updater's operation impossible to change in the future.

I've been feeling that we need some kind of API for update routine factories or something, like we did with the uninstallers, but I haven't figured out something good yet. It looks like we could really use a good use a solution though.

**Comment by JDGrimes** [comment]
The concern that keeps nagging me is how we'll handle this when the network version hasn't been set. I guess that partly depends on how things worked previously, and whether the network version was actually not set or not.

It looks like the old installables class would only use the network data if the installable was network activated, as far as determining whether to update. And it also handled updating the version in the DB, so it also only updated the per-site version for per-site active installables.

So in short, the version will not have previously been set at the network level.

As a result, all network update routines would be run on the first update, unless we set the network version in some way. We should set it on install, of course, but for existing installables, we could determine it by looping through all sites where they are installed, and getting the highest version from any of those sites (since that is the highest version of network updates that will have run, along with that update).

**Comment by JDGrimes** [comment]
Similar to the above, what about when an installable class is newly added to an entity? I guess that's no problem, the updates will run as expected since their version will be larger than an empty version.

Although we now run the installer in the installables app, which previously we did only for core. This means that when an installable first adds an installable object, it doesn't need to perform an update, it can just add an install routine, and that will get run by the installables app. This design may make things difficult, however, if the update was doing something that was only needed as an update (to correct a bug in the db, for example), and not during install. In that case, it would actually still have to be done as part of "install", because the update routine would not get run (when the install ran it would set the DB version, so updating wouldn't be triggered). This could potentially be problematic.

This change also means that just loading the plugin once accidentally will install it. So perhaps we should revisit this change in another ticket. Although this was already true for the plugin previously to #404, only now it is also true for all installables.

**Comment by JDGrimes** [comment]
> depending on the solution we may have to switch to each site only to update the version number.

Since we are keeping track of the network version number separately, this won't be necessary. The per-site versions don't need to be updated unless any updates are actually run.

**Comment by JDGrimes** [comment]
We've updated our legacy shim between the installables API and the un/installer API to pass in the network parameter to the un/installer's `update()` method. So we're now running per-site when not network-active, as we should be.

However, we're still running the network portion of uninstall each time we run on each site. So we need to decide whether to fix that or not.

I'm thinking that maybe it isn't worth addressing it, since the un/installer is now basically deprecated, and everything should be moving to the new API. As this bug has already existed for some time and hasn't caused any problems that we know of, I think that we'd be justified in just letting be for now.
